### ISSUE TITLE
UTF Import Issues

### ISSUE THREAD

**Issue by AtesComp** [OpenRefine/OpenRefine#4990]
Importing a file for a project needs to identify the file's UTF type to correctly ingest contents.  JavaScript and Java process strings natively as UTF-16.  An ASCII file read will convert to UTF-16.  But, an otherwise ASCII file can also be interpreted as UTF-8.  If sprinkled UTF-8 characters exist (such as NBSP), the UTF-8 characters are apparently imported as raw binary characters.  When imported to native UTF-16, these characters remain as UTF-8 sequences as there is no UTF type identifying the original file to perform a conversion.

This report is a WIP as variations of UTF processing may have unknown effects on an OpenRefine import.

### To Reproduce
1. Load a file with UTF-8 characters sprinkled in.  A late model MS Excel spreadsheet may end strings with a UTF-8 NBSP character as Windows 10 is UTF-8 by default.
2. Examine that cells end with an extra "space" character by highlighting the text.
3. Perform a GREL strip() on the text and examine that the "space" is not removed.

### Current Results
As per "To Reproduce" above.  UTF-8 characters are not processed properly into native UTF-16.

### Expected Behavior
UTF-8 and other UTF versions are processed properly into native UTF-16.  A solution is to ask the user for a UTF setting to process the import file.  Defaulting imports to UTF-8 may be acceptable for most instances.

### Screenshots
None

### Versions
 - Operating System: Linux, Ubuntu 21.04
 - Browser Version: Chrome
 - JRE or JDK Version: JDK 11
 - OpenRefine: OpenRefine 3.5.2

### Datasets
Example file to be posted,

### Additional context
None.

**Comment by wetneb** [comment]
Thanks! I think it would be useful to provide an example file for which this happens. I suspect this issue is specific to the Excel importer.

**Comment by thadguidry** [comment]
I think the best course of action is probably to **change the default charset for the XLSX importer to be UTF8**.  That eliminates at least a large majority of the problems that cause users to be surprised later on.  What do you think of just doing that @AtesComp in this issue?

**Comment by AtesComp** [comment]
I agree that that would generally be an acceptable base solution. It is hard to tell how the non-breaking space is getting into the text, but I suspect MS is doing it behind the scene as I've seen in MS Word documents--Word makes bad automatic decisions about adding non-breaking spaces.

Where is the affected code?  I need to get back on this with a sample file and test.

**Comment by thadguidry** [comment]
@AtesComp Start here: https://github.com/OpenRefine/OpenRefine/tree/master/main/src/com/google/refine/importers

**Comment by tfmorris** [comment]
I'm having a hard time understanding where this problem lies. Is it with XLS import or XLSX or both? What source program(s) ie versions of Excel or other spreadsheets? One or more files that demonstrate the problem would definitely be very useful (ie necessary).

I *think* that this should all be handled internally by Apache POI and/or the source program. For legacy files written by older versions of Excel, I suspect that the situation will be that you have to write the file using a code page which is expected by the receiving system and communicate that information out of band, just like in the olden days when folks mailed their floppies to each other.

In any case, a reproducer would help track it down.

@thadguidry The most common code page for legacy files will be CP1252. Modern versions of Excel should correctly set the code page when writing the files.

**Comment by thadguidry** [comment]
@tfmorris yes, modern Excel versions seem to be just fine.  Here is my test XLSX file for Excel 365 latest version.  It does not exhibit any problem and I could not get a NBSP to appear "automatically" at the end of text in cells unless I specifically typed in a NBSP char into the cell into Excel as shown in test file below.  I wonder if the problem might come up due to formula use?

[OpenRefine Excel XLSX Test Unicode.xlsx](https://github.com/OpenRefine/OpenRefine/files/9856700/OpenRefine.Excel.XLSX.Test.Unicode.xlsx)

![image](https://user-images.githubusercontent.com/986438/197670008-322bf663-651e-48ae-a6b0-26da283fc78c.png)

As far as @AtesComp suggestion of enhancing GREL `strip()` function to not only remove leading and trailing whitespace but also any leading and trailing U+00A0 (nbsp) chars, I would be +1👍 for that enhancement.

**Comment by tfmorris** [comment]
To clarify, the test case needed is one which reproduces the problem.

`strip()` is a synonym for `trim()` which already removes NBSP (and all other funky whitespace characters).

Note that there *is* an issue with line-based importers and NBSP which is described in #5246. There is a pending PR that fixes it.

**Comment by tfmorris** [comment]
I'm going to close this since it's been pending for almost 3 years. Please feel free to reopen if an example file which demonstrates the behavior becomes available

**Comment by AtesComp** [comment]
It is a very narrow corner case I don't think many other encountered. I'm good with the close.
